exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -114,30 +118,82 @@ impl JavaRuntimeProxy {
+            _ => Error::JniError.with_description(err.to_string()),
         }
     }
 
-    /// Executes closure `f` and handles any type of JNI errors from it.
+    /// Executes closure `f` and handles any type of JNI errors from it. Occurred
+    /// exceptions are cleared.
     ///
-    /// Any JNI errors are converted into `ExecutionError` with their descriptions, for JNI errors
-    /// like `JniErrorKind::JavaException` it gets (and clears) any exception that is currently
-    /// being thrown, `ExceptionHandlers::DEFAULT` is applied for such exceptions.
+    /// Used for runtime methods results of which do *not* depend on
+    /// implementation of the user service, but rather on implementation of
+    /// Java runtime and user input. E.g., `deploy_artifact`.
+    ///
+    /// - Any JNI errors are converted into `ExecutionError::Runtime(JniError)`.
+    /// - `IllegalArgumentException`s are converted into
+    ///   `ExecutionError::Runtime(IllegalArgument)`.
+    /// - Any other exceptions are converted into
+    ///   `ExecutionError::Runtime(JavaException)`.
     fn jni_call_default<F, R>(&self, f: F) -> Result<R, ExecutionError>
     where
         F: FnOnce(&JNIEnv) -> JniResult<R>,
     {
-        self.jni_call::<F, &ExceptionHandler, R>(&[], f)
+        self.jni_call::<F, &ExceptionHandler, R>(
+            &[(
+                &classes_refs::java_lang_illegal_argument_exception(),
+                ExceptionHandlers::ILLEGAL_ARGUMENT,
+            )],
+            f,
+        )
     }
 
-    /// Executes closure `f` and handles any type of JNI errors from it.
+    /// Executes closure `f` and handles any type of JNI errors from it. Occurred
+    /// exceptions are cleared.
     ///
-    /// Any JNI errors are converted into `ExecutionError` with their descriptions, for JNI errors
-    /// like `JniErrorKind::JavaException` it gets (and clears) any exception that is currently
-    /// being thrown, then exception is passed to corresponding `ExceptionHandler` according their
-    /// type and `exception_handlers` mapping.
-    /// `ExceptionHandlers::DEFAULT` is called in case of there is no any handlers or handlers are
-    /// not matched to exception type.
+    /// Used for runtime methods results of which depend on
+    /// implementation of the user service, such as `execute`,
+    /// `before_transactions` and `after_transactions`.
+    ///
+    /// - Any JNI errors are converted into `ExecutionError::Runtime(JniError)`.
+    /// - `IllegalArgumentException`s are converted into
+    ///   `ExecutionError::Runtime(IllegalArgument)`.
+    /// - `TransactionExecutionException`s are converted into
+    ///   `ExecutionError::Service`.
+    /// - `UnexpectedTransactionExecutionException`s are converted into
+    ///   `ExecutionError::Unexpected`.
+    /// - Any other exceptions are converted into
+    ///   `ExecutionError::Runtime(JavaException)`.
+    fn jni_call_transaction<F, R>(&self, f: F) -> Result<R, ExecutionError>
```

Comment:
I'd consider unit-testing these cases in native — they are easier to trigger than in Java ITs.

Reply:
Yes, sounds reasonable. Given that native ITs are disabled now, it doesn't make much sense implementing them in this PR. I'll create an issue with testing scenarios.

Reply:
Do we have an issue to fix existing ITs? I'd abstain from just piling issues and make the ITs pass to be able to add new tests promptly.